File: frontend/src/types/supplier.ts
/** Cf. backend/app/blueprints/suppliers/schemas.py */

export interface Supplier {
  id: string;
  name: string;
  contact_name: string | null;
  phone: string | null;
  email: string | null;
  address: string | null;
  is_active: boolean;
}

export interface SupplierWritePayload {
  name: string;
  contact_name?: string | null;
  phone?: string | null;
  email?: string | null;
  address?: string | null;
  is_active?: boolean;
}

export const RECEPTION_STATUSES = ["BROUILLON", "VALIDEE"] as const;
export type ReceptionStatus = (typeof RECEPTION_STATUSES)[number];

export interface ReceptionLineWritePayload {
  product_id: string;
  quantity: number;
  unit_purchase_price: string;
}

export interface ReceptionCreatePayload {
  supplier_id: string;
  branch_id: string;
  lines: ReceptionLineWritePayload[];
}

export interface ReceptionLine {
  id: string;
  product_id: string;
  product_sku: string;
  product_name: string;
  quantity: number;
  unit_purchase_price: string;
}

export interface Reception {
  id: string;
  reference: string;
  supplier_id: string;
  supplier_name: string;
  branch_id: string;
  branch_name: string;
  status: ReceptionStatus;
  received_at: string | null;
  created_at: string;
  lines: ReceptionLine[];
  total_amount: string;
}

export interface ReceptionListParams {
  branch_id?: string;
  supplier_id?: string;
  status?: ReceptionStatus;
}

File: backend/app/models/supplier.py
"""Modeles fournisseurs et receptions de marchandises."""
import enum

from app.extensions import db
from app.models.base import UUIDPrimaryKeyMixin, TimestampMixin


class Supplier(db.Model, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "suppliers"

    name = db.Column(db.String(120), nullable=False)
    contact_name = db.Column(db.String(120), nullable=True)
    phone = db.Column(db.String(32), nullable=True)
    email = db.Column(db.String(120), nullable=True)
    address = db.Column(db.String(255), nullable=True)
    is_active = db.Column(db.Boolean, nullable=False, default=True)


class ReceptionStatus(str, enum.Enum):
    BROUILLON = "BROUILLON"
    VALIDEE = "VALIDEE"


class SupplierReception(db.Model, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "supplier_receptions"

    reference = db.Column(db.String(32), unique=True, nullable=False)
    supplier_id = db.Column(db.String(36), db.ForeignKey("suppliers.id"), nullable=False)
    branch_id = db.Column(db.String(36), db.ForeignKey("branches.id"), nullable=False)
    status = db.Column(db.String(16), nullable=False, default=ReceptionStatus.BROUILLON.value)
    received_at = db.Column(db.DateTime, nullable=True)
    created_by_id = db.Column(db.String(36), db.ForeignKey("users.id"), nullable=False)

    supplier = db.relationship("Supplier", lazy="joined")
    branch = db.relationship("Branch", lazy="joined")
    lines = db.relationship(
        "SupplierReceptionLine", backref="reception", cascade="all, delete-orphan", lazy="joined"
    )


class SupplierReceptionLine(db.Model, UUIDPrimaryKeyMixin):
    __tablename__ = "supplier_reception_lines"

    reception_id = db.Column(
        db.String(36), db.ForeignKey("supplier_receptions.id"), nullable=False
    )
    product_id = db.Column(db.String(36), db.ForeignKey("products.id"), nullable=False)
    quantity = db.Column(db.Integer, nullable=False)
    unit_purchase_price = db.Column(db.Numeric(12, 2), nullable=False)

    product = db.relationship("Product", lazy="joined")

    __table_args__ = (
        db.CheckConstraint("quantity > 0", name="ck_supplier_reception_line_qty_positive"),
    )
